File: ground_station/lora/ground_station/libraries/kiss/kiss.cpp
#include "kiss.h"

bool Kiss::write(const uint8_t *data, uint32_t data_size, bool debug)
{
    std::list<uint8_t> frame;
    uint8_t b;
    /* Include two FEND */
    frame.push_back(KISS_FEND);
    /* Include content */
    for(unsigned int i = 0; i < data_size; i++) {
        b = data[i];
        if(b == KISS_FEND) {
            frame.push_back(KISS_FESC);
            frame.push_back(KISS_TFEND);
        } else if(b == KISS_FESC) {
            frame.push_back(KISS_FESC);
            frame.push_back(KISS_TFESC);
        } else {
            frame.push_back(b);
        }
    }
    /* Close the frame with FEND */
    frame.push_back(KISS_FEND);
    /* Send through the UART */
    uint8_t chunk[frame.size()];
    std::copy(frame.begin(), frame.end(), chunk);
    size_t written = Serial.write(chunk, frame.size());
    if(written != frame.size()) {
        if(debug == false) {
            LOG::err("Transmitted different amount of chars: %d != %d", written, frame.size());
        }
        return false;
    }
    return true;
}

uint32_t Kiss::read(uint8_t *data, uint32_t data_size)
{
    uint8_t b;
    bool started_frame = false;
    bool finished_frame = false;
    bool fesc_found = false;
    std::list<uint8_t> frame;
    double start_time;

    if(Serial.available() > 0) {
        start_time = millis();
        while(finished_frame == false) {
            if(Serial.available() > 0) {
                b = (uint8_t)Serial.read();
                /* There is data to process */
                if((b & 0xFF) == KISS_FEND && started_frame == false) {
                    started_frame = true;
                } else if((b & 0xFF) == KISS_FEND && finished_frame == false) {
                    finished_frame = true;
                } else if((b & 0xFF) == KISS_FESC && fesc_found == false) {
                    fesc_found = true;
                } else if((b & 0xFF) == KISS_TFEND && fesc_found == true) {
                    frame.push_back(KISS_FEND);
                    fesc_found = false;
                } else if((b & 0xFF) == KISS_TFESC && fesc_found == true) {
                    frame.push_back(KISS_FESC);
                    fesc_found = false;
                } else if(started_frame == true){
                    frame.push_back(b);
                }
                if(started_frame == true && finished_frame == false) {
                    start_time = millis();
                }
            } else if(millis() >= start_time + 10) {    // waiting 10 ms to receive a byte
                //LOG::warn("Started KISS frame, but no char received; cleanning frame.");
                return 0;
            }
        }
        std::copy(frame.begin(), frame.end(), data);
    }
    return frame.size();
}

uint32_t Kiss::available()
{
    return Serial.available();
}

File: ground_station/python/kiss.py
#! /usr/bin/env python

import serial
import time

KISS_FEND = b'\xc0'
KISS_FESC = b'\xdb'
KISS_TFESC = b'\xdc'
KISS_TFEND = b'\xdd'

current_milli_time = lambda: int(round(time.time() * 1000))

class Kiss:
    def __init__(self, port, speed, bytesize, parity, stopbits, timeout):
        self.port = serial.Serial(port, speed, bytesize, parity, stopbits, timeout)

    def is_open(self):
        return self.port.is_open

    def read(self):
        started_frame = False
        finished_frame = False
        fesc_found = False
        frame = b''
        if(self.port.in_waiting > 0):
            start_time = current_milli_time();
            while(finished_frame == False):
                if(self.port.in_waiting > 0):
                    b = self.port.read()
                    # There is data to process
                    if(b == KISS_FEND and started_frame == False):
                        started_frame = True;
                    elif(b == KISS_FEND and finished_frame == False):
                        finished_frame = True;
                    elif(b == KISS_FESC and fesc_found == False):
                        fesc_found = True;
                    elif(b == KISS_TFEND and fesc_found == True):
                        frame += KISS_FEND
                        fesc_found = False
                    elif(b == KISS_TFESC and fesc_found == True):
                        frame += KISS_FESC
                        fesc_found = False;
                    elif(started_frame == True):
                        frame += b
                elif(current_milli_time() >= start_time + 10):   # waiting 10 ms to receive a byte
                    print("[KISS] Frame started but not finished")
                    return b'';
                time.sleep(0.001)
        return frame;

    def write(self, data):
        # Include first FEND
        frame = KISS_FEND
        # Include content
        for i in range(1, len(data) + 1):
            b = data[i-1:i]
            if(b == KISS_FEND):
                frame += KISS_FESC
                frame += KISS_TFEND
            elif(b == KISS_FESC):
                frame += KISS_FESC
                frame += KISS_TFESC
            else:
                frame += b
        # Close the frame with FEND
        frame += KISS_FEND
        # Send through the UART
        written = self.port.write(frame);
        if(written != len(frame)):
            return False;
        return True;

    def close(self):
        self.port.close()
